src/shogun/lib/SGMatrix.cpp

```diff
@@ -227,6 +227,22 @@ void SGMatrix<T>::zero()
 	set_const(static_cast<T>(0));
 }
 
+template <class T>
+void SGMatrix<T>::random(T min_value, T max_value)
+{
+	for (index_t idx = 0; idx < num_rows * num_cols; ++idx)
```

Comment:
pls, instead of a loop, use the `CRandom::fill_array*` functions that in some situations can be much much faster

Reply:
I was making it consistent with `SGVector` but sure - yes I noticed it uses SIMD instructions.
I think `SGVector::random` should be changed too accordingly - not sure whether should be done in this or a different PR.

Reply:
1. yes you are exactly right, this needs template specialization, and a loop fallback is fine imo
Let's do it for SGMatrix::random in this PR
2. I suggest we keep the interface without arguments, i.e. (`SGMatrix::random()` and `SGMatrix::random_normal`), and then the caller is responsible for scaling things to the appropriate bounds, using linalg

very nicely thought trough! :)

Reply:
Thanks, but I'd like to clarify (2) again. I get the `SGMatrix::random()` function which would just fill the array with a uniform random distribution(using `fill_array*`). For `random_normal` I would need to further apply something like the [Box-Muller transform](https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform) right? - I could try to do that using the `shogun::linalg` ops though not sure if its feasible with those alone.

Reply:
@saatvikshah1994 
we have two options here:
1. Make a loop over the CRandom::random_normal insice `SGMatrix::random (since CRandom doesn't support SIMD random normal generation 
2. Implement a random_normal_array which
 a. Has a default implementation with a loop for now
 b. First uses fill_array and then re-uses the box muller trainsform code that is already present (would need some refactoring)

I suggest we keep things simple for now:
* offer `CRandom::random_normal_array` and use a loop for now
* if you want: make another PR once this is merged where you do option 2b from above

We definitely only want `SGMatrix::random_normal` to be a wrapper, i.e. no transforms or other complicated things in there.

Reply:
@karlnapf thanks this makes sense.
I had one suggestion though - I feel that the entire `random()` and `random_normal()` impl should go in its own separate (issue + PR combo). Reason being that (1) the implementation is less related to the immediate issue of minimizing the test redundancies and (2) the changes should ideally be propagated to `SGVector` as well - overall then a significant portion of this PR would mix up two disparate issues. For this PR, I could just use the original `NeuralLayerFixture::create_rand_matrix` which would then be replaced in the second PR. 

I'm up for whichever way you lean towards.